### DISCUSSION TITLE
Deserialize error

### DISCUSSION THREAD

**Discussion by dengzhq** [tokio-rs/axum#1359] [Q&A]
### Version
0.5.15

### Platform
64-bit (Windows)

### Crates
the crates where use `serde_html_form::from_str()` method

### Description
For code reuse, I wrote this code with the help of serde's flatten attr,
```
#[derive(Debug, serde::Deserialize)]
pub struct RightPageParam {
    pub name: String,
    pub id: u32,
    #[serde(flatten)]
    pub abstract_query_param: AbstractQueryParam,
}
#[derive(serde::Deserialize, Debug)]
pub struct AbstractQueryParam {
    current: Option<u32>,
    size: Option<u32>,
}
fn main() {
    let path_and_query: PathAndQuery = "/getPage?name=12&id=5&current=25&size=10".parse().unwrap();
    let query = path_and_query.query().unwrap();
    println!("query = {:?}", query);

    let value: RightPageParam = serde_html_form::from_str(query).map_err(FailedToDeserializeQueryString::__private_new::<RightPageParam, serde_html_form::de::Error>).unwrap();
    println!("value = {:?}", value);
}

```

but when I try to use it in a `get` method， it throws an exception
```
called `Result::unwrap()` on an `Err` value: FailedToDeserializeQueryString { error: Error { inner: Error("invalid type: string \"25\", expected u32") } }
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: FailedToDeserializeQueryString { error: Error { inner: Error("invalid type: string \"25\", expected u32") } }',
```
I checked the source code, the exception comes from 
```rust
#[async_trait]
impl<T, S> FromRequestParts<S> for Query<T>
where
    T: DeserializeOwned,
    S: Send + Sync,
{
    type Rejection = QueryRejection;

    async fn from_request_parts(parts: &mut Parts, _state: &S) -> Result<Self, Self::Rejection> {
        let query = parts.uri.query().unwrap_or_default();
        let value = serde_html_form::from_str(query)
            .map_err(FailedToDeserializeQueryString::__private_new)?;
        Ok(Query(value))
    }
}
```
Is it a bug? or is there a better way to organize the code?

**Comment by davidpdrsn** [comment]
The code you posted doesn't call anything in axum. You don't use `Query` in your example.

I'm not quite sure what the problem is. Maybe something with `#[serde(flatten)]` in serde_html_form.

**Reply by dengzhq** [reply]
sorry，I just used the minimal code from the question, my original code is like this
```
Router::new() .route("/getPage", get(get_page)) ;
#[axum_macros::debug_handler]
pub async fn get_page(Query(param): Query<PageParam>) -> BizJsonResult<XpLarkPage<String>> {
    let records = XpLarkPage::new(vec!["chenhm".to_owned(),"dengzhq".to_owned()], 2, param.abstract_query_param.size(), param.abstract_query_param.current());
    result_util::success_json(records)
}
```
the PageParam is the same RightPageParam .
when i run:http://localhost:8082/getPage?name=12&id=5&current=25&size=10。the server return:
```
Failed to deserialize query string: invalid type: string "25", expected u32
```
Is this more helpful?

**Reply by davidpdrsn** [reply]
I see. There isn't anything we can do about it in axum though. When you look at how `Query` is implemented you see it just calls into serde_html_form (or serde_url_encoded)

So either it's a bug in their implementation or it requires you to annotate your field with a [serde field attribute](https://serde.rs/field-attrs.html) to make the parsing more flexible.

**Reply by dengzhq** [reply]
@jplatte hey，as the crate `serde_html_form`  author,what do you think? Is it a `serde` problem or  `serde_html_form` problem

**Reply by jplatte** [reply]
This is a problem of how `serde(flatten)` interacts with some parts of serde's overall way of working. It is unfortunately impossible to fix in either `serde` itself or `serde_html_form` without breaking other things.

**Reply by jplatte** [reply]
If you look for issues in serde containing the word `flatten`, you will find that there is a lot of them, all without any actual movement. The oldest one is https://github.com/serde-rs/serde/issues/1183 and kind of related.

My recommendation is to avoid `serde(flatten)` whenever possible (it also hurts performance) .

**Reply by dengzhq** [reply]
oh i see, Thanks for all your answers
